src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.OSX.cs

```diff
@@ -100,49 +64,113 @@ private void StopRaisingEvents()
-        private CFRunLoopRef _watcherRunLoop = IntPtr.Zero;
+            // The EventStream to listen for events on
+            private SafeEventStreamHandle _eventStream;
 
-        // Since all variables are IntPtr, Strings, or change references, using this in our locks
-        private readonly object _syncLock = new object();
+            // A reference to the RunLoop that we can use to start or stop a Watcher
+            private CFRunLoopRef _watcherRunLoop;
 
-        // Callback delegate for the EventStream events 
-        private Interop.EventStream.FSEventStreamCallback _callback = null;
+            // Callback delegate for the EventStream events 
+            private Interop.EventStream.FSEventStreamCallback _callback;
 
-        // Use an event to try to prevent StartRaisingEvents from returning before the
-        // RunLoop actually begins. This will mitigate a race condition where the watcher
-        // thread hasn't completed initialization and stop is called before the RunLoop even starts.
-        private readonly AutoResetEvent _runLoopStartedEvent = new AutoResetEvent(false);
+            // Token to monitor for cancellation requests, upon which processing is stopped and all
+            // state is cleaned up.
+            private readonly CancellationToken _cancellationToken;
 
-        private void CreateStreamAndStartWatcher()
-        {
-            Debug.Assert(_eventStream.IsInvalid);
-            Debug.Assert(_watcherRunLoop == IntPtr.Zero);
-            Debug.Assert(_callback == null);
+            // Calling RunLoopStop multiple times SegFaults so protect the call to it
+            private bool _stopping;
+
+            internal RunningInstance(
+                FileSystemWatcher watcher,
+                string directory,
+                bool includeChildren,
+                Interop.EventStream.FSEventStreamEventFlags filter,
+                CancellationToken cancelToken)
+            {
+                Debug.Assert(string.IsNullOrEmpty(directory) == false);
+                Debug.Assert(!cancelToken.IsCancellationRequested);
+
+                _weakWatcher = new WeakReference<FileSystemWatcher>(watcher);
+                _fullDirectory = System.IO.Path.GetFullPath(directory);
+                _includeChildren = includeChildren;
+                _filterFlags = filter;
+                _cancellationToken = cancelToken;
+                _cancellationToken.Register(obj => ((RunningInstance)obj).CancellationCallback(), this);
+                _stopping = false;
+            }
 
-            // Make sure we only do this if there is a valid directory
-            if (String.IsNullOrEmpty(_directory) == false)
+            private void CancellationCallback()
             {
-                _fullDirectory = System.IO.Path.GetFullPath(_directory);
+                if (!_stopping)
+                {
+                    _stopping = true;
```

Comment:
Rather than having another field / flag, could you just do:
```C#
IntPtr loop = _watcherRunLoop;
if (loop != IntPtr.Zero)
{
    _watcherRunLoop = IntPtr.Zero;
    Interop.RunLoop.CFRunLoopStop(loop);
}
```
?